src/variant_iter.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::iter::Iterator;
+
+use translate::*;
+use variant::Variant;
+
+glib_wrapper! {
+    /// Iterator over items in a variant.
+    pub struct VariantIter(Boxed<glib_sys::GVariantIter>);
+
+    match fn {
+        copy => |ptr| glib_sys::g_variant_iter_copy(mut_override(ptr)),
+        free => |ptr| glib_sys::g_variant_iter_free(ptr),
+    }
+}
+
+impl Iterator for VariantIter {
+    type Item = Variant;
+
+    fn next(&mut self) -> Option<Variant> {
+        unsafe {
+            let value = glib_sys::g_variant_iter_next_value(mut_override(self.to_glib_none().0));
+            if !value.is_null() {
+                Some(from_glib_full(value))
+            } else {
+                None
+            }
+        }
+    }
```

Comment:
You probably want to implement `Iterator::size_hint()` and `ExactSizeIterator` here around `g_variant_iter_n_children()`. That would make things like `Iterator::collect()` work better :)

Note that the Rust functions are about the remaining elements, while the `GVariantIter` function is about the overall elements.

Reply:
Yeah, I thought about that.

> Note that the Rust functions are about the remaining elements, while the GVariantIter function is about the overall elements.

This part is a bit of a pain, since it means it would have to maintain a count of how many elements it has gone through so far. I guess the overhead of that wouldn't be too significant. But I suppose `glib_wrapper!` wouldn't support adding an additional member to the wrapper struct.

Reply:
Indeed, that sucks :) Maybe instead we can implement an iterator ourselves around counting? Instead of using the C `GVariantIter`. That seems equally simple and there doesn't seem to be any advantage in using the C iterator, it just adds another heap allocation.

Reply:
Hm. I wasn't sure if it the C implementation had any advantages over attempting to reimplement it (if variants used linked lists for instance, it could iterate in a more efficient manner).

But looking at the glib source code, it seems like `g_variant_iter_next_value()` is just implemented with `g_variant_get_child_value()`, and stores a `i` and `n` with the length and position in the container, and a reference to the `Variant`.

I don't see any reason that couldn't be reimplemented in Rust without an extra heap allocation.